File: ledger-models-python/fintekkers/wrappers/models/security/mortgage_backed_security.py
"""MortgageBackedSecurity wrapper.

Specialises BondSecurity for agency mortgage-backed securities (MBS). The
shared bond fields (coupon, dates, face value) live on
SecurityProto.bond_details; MBS-specific extras (pool number, agency,
WAC, WAM, pass-through rate, current factor, original face value,
current UPB, PSA speed) live on SecurityProto.mbs_extension. The two
sub-messages coexist - neither replaces the other.
"""

from datetime import date
from decimal import Decimal
from typing import Optional

from fintekkers.models.security.security_pb2 import (
    SecurityProto,
    BondDetailsProto,
    MbsExtensionProto,
)
from fintekkers.models.security.product_type_pb2 import ProductTypeProto
from fintekkers.models.security.bond.agency_pb2 import AgencyProto
from fintekkers.wrappers.models.security.bond_security import (
    BondSecurity,
    _decimal_value,
    _local_date,
)


class MortgageBackedSecurity(BondSecurity):
    """Wraps a SecurityProto with product_type=MORTGAGE_BACKED."""

    def get_pool_number(self) -> str:
        self._assert_not_link("pool_number")
        if not self.proto.HasField("mbs_extension"):
            return ""
        return self.proto.mbs_extension.pool_number

    def get_agency(self) -> AgencyProto:
        self._assert_not_link("agency")
        if not self.proto.HasField("mbs_extension"):
            return AgencyProto.AGENCY_UNKNOWN
        return self.proto.mbs_extension.agency

    def get_wac(self) -> Optional[Decimal]:
        self._assert_not_link("wac")
        if not self.proto.HasField("mbs_extension"):
            return None
        ext = self.proto.mbs_extension
        if not ext.HasField("wac"):
            return None
        value = ext.wac.arbitrary_precision_value
        if value == "":
            return None
        return Decimal(value)

    def get_wam(self) -> int:
        self._assert_not_link("wam")
        if not self.proto.HasField("mbs_extension"):
            return 0
        return self.proto.mbs_extension.wam

    def get_pass_through_rate(self) -> Optional[Decimal]:
        self._assert_not_link("pass_through_rate")
        if not self.proto.HasField("mbs_extension"):
            return None
        ext = self.proto.mbs_extension
        if not ext.HasField("pass_through_rate"):
            return None
        value = ext.pass_through_rate.arbitrary_precision_value
        if value == "":
            return None
        return Decimal(value)

    def get_current_factor(self) -> Optional[Decimal]:
        self._assert_not_link("current_factor")
        if not self.proto.HasField("mbs_extension"):
            return None
        ext = self.proto.mbs_extension
        if not ext.HasField("current_factor"):
            return None
        value = ext.current_factor.arbitrary_precision_value
        if value == "":
            return None
        return Decimal(value)

    def get_original_face_value(self) -> Optional[Decimal]:
        self._assert_not_link("original_face_value")
        if not self.proto.HasField("mbs_extension"):
            return None
        ext = self.proto.mbs_extension
        if not ext.HasField("original_face_value"):
            return None
        value = ext.original_face_value.arbitrary_precision_value
        if value == "":
            return None
        return Decimal(value)

    def get_current_upb(self) -> Optional[Decimal]:
        self._assert_not_link("current_upb")
        if not self.proto.HasField("mbs_extension"):
            return None
        ext = self.proto.mbs_extension
        if not ext.HasField("current_upb"):
            return None
        value = ext.current_upb.arbitrary_precision_value
        if value == "":
            return None
        return Decimal(value)

    def get_psa_speed(self) -> Optional[Decimal]:
        self._assert_not_link("psa_speed")
        if not self.proto.HasField("mbs_extension"):
            return None
        ext = self.proto.mbs_extension
        if not ext.HasField("psa_speed"):
            return None
        value = ext.psa_speed.arbitrary_precision_value
        if value == "":
            return None
        return Decimal(value)

    @staticmethod
    def from_pricer_inputs(
        face_value: Decimal,
        coupon_rate: Decimal,
        coupon_type: int,
        coupon_frequency: int,
        issue_date: date,
        maturity_date: date,
        pool_number: str,
        agency: int,
        wac: Decimal,
        wam: int,
        pass_through_rate: Decimal,
        current_factor: Decimal,
        original_face_value: Decimal,
        current_upb: Decimal,
        psa_speed: Decimal,
    ) -> SecurityProto:
        """Builds an MBS SecurityProto with both bond_details and
        mbs_extension populated, and product_type=MORTGAGE_BACKED.
        """
        bond_details = BondDetailsProto(
            coupon_rate=_decimal_value(coupon_rate),
            coupon_type=coupon_type,
            coupon_frequency=coupon_frequency,
            face_value=_decimal_value(face_value),
            issue_date=_local_date(issue_date),
            maturity_date=_local_date(maturity_date),
        )
        mbs_extension = MbsExtensionProto(
            pool_number=pool_number,
            agency=agency,
            wac=_decimal_value(wac),
            wam=wam,
            pass_through_rate=_decimal_value(pass_through_rate),
            current_factor=_decimal_value(current_factor),
            original_face_value=_decimal_value(original_face_value),
            current_upb=_decimal_value(current_upb),
            psa_speed=_decimal_value(psa_speed),
        )
        return SecurityProto(
            product_type=ProductTypeProto.MORTGAGE_BACKED,
            bond_details=bond_details,
            mbs_extension=mbs_extension,
        )

File: ledger-models-rust/fintekkers/wrappers/models/mortgage_backed_security.rs
//! Typed wrapper around an Agency-MBS-shaped `SecurityProto`.
//!
//! Agency MBS pass-throughs carry shared bond fields on `bond_details` plus
//! MBS-specific extras (pool number, agency, WAC, WAM, pass-through rate,
//! current factor, original face, current UPB, PSA speed) on `mbs_extension`.
//! This wrapper exposes the MBS-specific accessors and a builder
//! (`from_pricer_inputs`) that produces a fully-populated `SecurityProto`
//! for the pricer.

use chrono::{Datelike, NaiveDate};
use rust_decimal::Decimal;

use crate::fintekkers::models::security::bond::AgencyProto;
use crate::fintekkers::models::security::{
    BondDetailsProto, CouponFrequencyProto, CouponTypeProto, MbsExtensionProto, ProductTypeProto,
    SecurityProto,
};
use crate::fintekkers::models::util::{DecimalValueProto, LocalDateProto};
use crate::fintekkers::wrappers::models::utils::errors::Error;
use crate::fintekkers::wrappers::models::utils::serialization::ProtoSerializationUtil;

pub struct MortgageBackedSecurity {
    pub proto: SecurityProto,
}

impl MortgageBackedSecurity {
    pub fn from_proto(proto: SecurityProto) -> Result<Self, Error> {
        let pt = ProductTypeProto::from_i32(proto.product_type)
            .unwrap_or(ProductTypeProto::ProductTypeUnknown);
        if pt == ProductTypeProto::MortgageBacked {
            Ok(MortgageBackedSecurity { proto })
        } else {
            Err(Error::NotABondSecurity)
        }
    }

    pub fn pool_number(&self) -> &str {
        self.proto
            .mbs_extension
            .as_ref()
            .map(|m| m.pool_number.as_str())
            .unwrap_or("")
    }

    pub fn agency(&self) -> AgencyProto {
        self.proto
            .mbs_extension
            .as_ref()
            .and_then(|m| AgencyProto::from_i32(m.agency))
            .unwrap_or(AgencyProto::AgencyUnknown)
    }

    pub fn wac(&self) -> Option<Decimal> {
        self.proto
            .mbs_extension
            .as_ref()
            .and_then(|m| m.wac.as_ref())
            .and_then(|d| ProtoSerializationUtil::deserialize_decimal(d).ok())
    }

    pub fn wam(&self) -> i32 {
        self.proto
            .mbs_extension
            .as_ref()
            .map(|m| m.wam)
            .unwrap_or(0)
    }

    pub fn pass_through_rate(&self) -> Option<Decimal> {
        self.proto
            .mbs_extension
            .as_ref()
            .and_then(|m| m.pass_through_rate.as_ref())
            .and_then(|d| ProtoSerializationUtil::deserialize_decimal(d).ok())
    }

    pub fn current_factor(&self) -> Option<Decimal> {
        self.proto
            .mbs_extension
            .as_ref()
            .and_then(|m| m.current_factor.as_ref())
            .and_then(|d| ProtoSerializationUtil::deserialize_decimal(d).ok())
    }

    pub fn original_face_value(&self) -> Option<Decimal> {
        self.proto
            .mbs_extension
            .as_ref()
            .and_then(|m| m.original_face_value.as_ref())
            .and_then(|d| ProtoSerializationUtil::deserialize_decimal(d).ok())
    }

    pub fn current_upb(&self) -> Option<Decimal> {
        self.proto
            .mbs_extension
            .as_ref()
            .and_then(|m| m.current_upb.as_ref())
            .and_then(|d| ProtoSerializationUtil::deserialize_decimal(d).ok())
    }

    pub fn psa_speed(&self) -> Option<Decimal> {
        self.proto
            .mbs_extension
            .as_ref()
            .and_then(|m| m.psa_speed.as_ref())
            .and_then(|d| ProtoSerializationUtil::deserialize_decimal(d).ok())
    }

    pub fn from_pricer_inputs(
        face_value: Decimal,
        coupon_rate: Decimal,
        coupon_type: CouponTypeProto,
        coupon_frequency: CouponFrequencyProto,
        issue_date: NaiveDate,
        maturity_date: NaiveDate,
        pool_number: String,
        agency: AgencyProto,
        wac: Decimal,
        wam: i32,
        pass_through_rate: Decimal,
        current_factor: Decimal,
        original_face_value: Decimal,
        current_upb: Decimal,
        psa_speed: Decimal,
    ) -> SecurityProto {
        SecurityProto {
            product_type: ProductTypeProto::MortgageBacked as i32,
            bond_details: Some(BondDetailsProto {
                coupon_rate: Some(decimal_proto(coupon_rate)),
                coupon_type: coupon_type as i32,
                coupon_frequency: coupon_frequency as i32,
                face_value: Some(decimal_proto(face_value)),
                issue_date: Some(date_proto(issue_date)),
                dated_date: None,
                maturity_date: Some(date_proto(maturity_date)),
                issuance_info: vec![],
            }),
            mbs_extension: Some(MbsExtensionProto {
                pool_number,
                agency: agency as i32,
                wac: Some(decimal_proto(wac)),
                wam,
                pass_through_rate: Some(decimal_proto(pass_through_rate)),
                current_factor: Some(decimal_proto(current_factor)),
                original_face_value: Some(decimal_proto(original_face_value)),
                current_upb: Some(decimal_proto(current_upb)),
                psa_speed: Some(decimal_proto(psa_speed)),
            }),
            ..Default::default()
        }
    }
}

fn decimal_proto(d: Decimal) -> DecimalValueProto {
    DecimalValueProto {
        arbitrary_precision_value: d.to_string(),
    }
}

fn date_proto(d: NaiveDate) -> LocalDateProto {
    LocalDateProto {
        year: d.year() as u32,
        month: d.month(),
        day: d.day(),
    }
}
